compiler/rustc_target/src/spec/hexagon_unknown_linux_musl.rs

```diff
@@ -9,7 +9,11 @@ pub fn target() -> Target {
 
     base.crt_static_default = false;
     base.has_rpath = true;
-    base.linker_flavor = LinkerFlavor::Unix(Cc::Yes);
+    base.linker_flavor = LinkerFlavor::Gnu(Cc::No, Lld::Yes);
```

Comment:
Do you need rustc to pass `-fuse-ld=lld` to clang in this case?
Or will clang automatically use lld for this target without any explicit options?

Reply:
> maybe I should instead change the expected results for the test case?

IIRC usually these tests rely on the values set to be different from the default values, to emulate what happens with a real CLI session. So, when changing the defaults, the test is changed to set a value that is detectable.

Reply:
> > maybe I should instead change the expected results for the test case?
> 
> IIRC usually these tests rely on the values set to be different from the default values, to emulate what happens with a real CLI session. So, when changing the defaults, the test is changed to set a value that is detectable.

The [test in question is "just" a roundtrip to JSON AFAICT](https://github.com/rust-lang/rust/blob/9b72cc9abf35f82e90cdc5fc54287e87aa2d3aeb/compiler/rustc_target/src/spec/tests/tests_impl.rs#L5-L10).  So I'm not quite clear on how changing from `Cc::No` to `Cc::Yes` in the `LinkerFlavor::Gnu` causes the test to fail, unless something about the JSON serialization/deserialization doesn't work for this scenario.  I'll take a closer look.

Reply:
>`clang --target=hexagon-unknown-linux-musl` implies `-fuse-ld=lld`.

In this case it should be fine to use `Cc::Yes, Lld::No`.
`Lld::No` just means that rustc won't pass any additional options to the C compiler.

Reply:
It would still be good to figure out why the roundtrip test fails with `Cc::Yes, Lld::Yes`.

Reply:
I believe the normalization step causes this mismatch: https://github.com/rust-lang/rust/blob/e7f9f48d767c1e17e125d83e3102d568f87b868d/compiler/rustc_target/src/spec/mod.rs#L2089-L2091

Reply:
What's the implication here?  Sorry to be a bit naive but how does the `json` representation relate to the ones specified in code?  Is there a goal/direction moving towards one or the other?  Or is it expected that we want to continue using both?

Reply:
I don't _think_ the round-trip fails because of the normalization step linked above because you don't have link args.

A `LinkerFlavor::Gnu(Cc::Yes, Lld::Yes)` will be split into three json fields:
- `linker_flavor_json` will be `Gcc`
- `lld_flavor_json`: `Ld`
- `linker_is_gnu_json`: `true`

to then be reconstructed in `LinkerFlavor::from_cli_json`, where the above triplet will be matched to `LinkerFlavor::Gnu(Cc::Yes, Lld::No)`, triggering the test failure. 

---

My understanding is that json targets are unstable, and therefore we could be using the unstable `-Clinker-flavor`s there without breaking anything? 

They will roundtrip, and the legacy stable values will still work for the existing/common cases (the ones that are not `(Cc::Yes, Lld::Yes)`). Even if we actually want to do that in the near future, it should be unrelated here, and we wouldn't need to do that for the sole purpose of the json round-trip.

That is, we could expand `LinkerFlavor::to_cli` to emit the corresponding modern `LinkerFlavorCli`s (and possibly stop emitting the legacy ones altogether), e.g. adding a `LinkerFlavor::Gnu(Cc::Yes, Lld::Yes) => LinkerFlavorCli::Gnu(Cc::Yes, Lld::Yes)` arm there should make roundtrip possible for the minimal case here. This would likely impact some of the CLI diagnostics: flavor compatibility is computed via these roundtrips, and that may need to be tweaked.

What do you think @petrochenkov ?

(We could imagine serializing the flavor to a single json field now that the linker flavor is precise enough, *and* still accept deserializing the old triplet of json fields for backwards compatibility).